### PR TITLE
Disable bloom filtering by default.

### PR DESCRIPTION
**Pull request by TheBlueMatt**
BIP 37 bloom filters have been well-known to be a significant DoS
target for some time. However, in order to provide continuity for
SPV clients relying on it, the NODE_BLOOM service flag was added,
and left as a default, to ensure sufficient nodes exist with such a
flag.

NODE_BLOOM is, at this point, well-established and, as long as
there exist 0.18 nodes with default config (which I'd anticipate
will be true for many years), will be available from some peers. By
that time, the continued slowdown of BIP 37-based filtering will
likely have rendered it useless (though this is already largely the
case). Further, BIP 37 was deliberately never updated to support
witness-based filtering as newer wallets are expected to migrate to
some yet-to-be-network-exposed filters.

### CODE REVIEW

**Comment by kallewoof** [comment]
Concept ACK

File: src/validation.h [hunk 1]
```diff
@@ -127,7 +127,7 @@ static const unsigned int MAX_BLOCKS_TO_ANNOUNCE = 8;
 /** Maximum number of unconnecting headers announcements before DoS score */
 static const int MAX_UNCONNECTING_HEADERS = 10;
 
-static const bool DEFAULT_PEERBLOOMFILTERS = true;
+static const bool DEFAULT_PEERBLOOMFILTERS = false;
```
Comment (MarcoFalke) [inline src/validation.h]: This has nothing to do with validation, please move it to net or something.

**Comment by MarcoFalke** [comment]
Also needs release notes and a post to the mailing list

File: src/validation.h (hunk 1 above)
Comment (PastaPastaPasta) [reply src/validation.h]: agree on that probably being better in different pr

Comment (TheBlueMatt) [reply src/validation.h]: No, we already have more than enough volume of useless "move code around" PRs. This is precisely the kind of thing that should be in the same PR in a separate commit.

Comment (laanwj) [reply src/validation.h]: let's not shedpaint this but imo it's fine to ombine this small code structure improvement with the change

**Comment by laanwj** [comment]
concept and code-review ACK, only point of discussion is what release to set as milestone here

**Comment by TheBlueMatt** [comment]
Linter appears broken.

**Comment by TheBlueMatt** [comment]
I'd presume next-major. No need to rush it into a backport, let alone to ensure there is always gonna be reasonable coverage of NODE_BLOOM nodes available.

**Comment by jtimon** [comment]
utACK f27309f55c4fa2b115525d72abb280757a568709

**Comment by TheBlueMatt** [comment]
> Also needs release notes and a post to the mailing list

This is not a protocol change, only a default argument change. No need to do a mailing list post. I added a release notes entry.

**Comment by luke-jr** [comment]
utACK, without affirmation of future plans mentioned in PR description.

**Comment by jnewbery** [comment]
ACK 048e0dcaea555e11bbdc8c56de0b9a523e42d351

I've read the code and it seems reasonable to me. It's surprising that bloom filters aren't tested by the functional tests.

This line could be removed: https://github.com/bitcoin/bitcoin/blob/8a503a6c6dd419921373f45d7aa6f1787c9b8884/test/functional/p2p_mempool.py#L20 since `-peerbloomfilters` now defaults to false, but no harm in leaving it.

**Comment by NicolasDorier** [comment]
NACK, @TheBlueMatt bloom filter should always be enabled for white listed peers, as this is the most effective way to sync a lightweight client from your own full node without the need of additional middleware like electrum.

Even BIP157 come short for this usecase as it needlessly waste bandwidth.

**Comment by TheBlueMatt** [comment]
@NicolasDorier That seems like an orthogonal issue. There is currently no code which enables it by default for whitelisted peers, so feel free to open a separate PR that does that.

**Comment by NicolasDorier** [comment]
I don't believe it is completely orthogonal.

Changing this default will break existing software which rely on the old default, at least we can limit damage to non-whitelisted peers only.

But yeh, anyway Concept ACK, not worth wasting ink on this, would just prefer if it was done here as well.

**Comment by TheBlueMatt** [comment]
0.19 is a ways off, no big deal either way :p

File: test/functional/test_framework/messages.py [hunk 2]
```diff
@@ -44,7 +44,7 @@
 
 NODE_NETWORK = (1 << 0)
 # NODE_GETUTXO = (1 << 1)
-NODE_BLOOM = (1 << 2)
```
Comment (NicolasDorier) [inline test/functional/test_framework/messages.py]: Do not disable this, I will make a test later for allowing NODE_BLOOM services to whitelisted peers.

Comment (jnewbery) [reply test/functional/test_framework/messages.py]: No need to invalidate reviews. Just uncomment it in your PR.

Comment (jnewbery) [reply test/functional/test_framework/messages.py]: (also linting tools like vulture may flag an error if this constant is defined but not used)

**Comment by TheBlueMatt** [comment]
Exactly, if you don't comment that line the linters decide that you've added a Clear Bug and must be shot.

**Comment by Empact** [comment]
Concept ACK

File: test/functional/test_framework/messages.py (hunk 2 above)
Comment (luke-jr) [reply test/functional/test_framework/messages.py]: That sounds like a bug in the linting tool IMO.

File: doc/release-notes/release-notes-16152.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Miscellaneous CLI Changes
```
Comment (MarcoFalke) [inline doc/release-notes/release-notes-16152.md]: This is not a CLI change, but a change of default p2p behavior. (You remove the `NODE_BLOOM` service bit by default)

**Review by maflcko — APPROVED** [review]
ACK, but could squash the commits, since the code changes are too trivial to justify three separate commits.

**Comment by ajtowns** [comment]
ACK 048e0dcaea555e11bbdc8c56de0b9a523e42d351 ; light code review, checked it compiled.

> NODE_BLOOM is, at this point, well-established and, as long as
> there exist 0.18 nodes with default config (which I'd anticipate
> will be true for many years), will be available from some peers.

I don't think this necessarily holds up -- if percentage `p` of nodes disable bloom filters (by upgrading to 0.19), while bloom users remain constant, then the load on the remaining nodes goes up by a factor of `1/(1-p)`; at some threshold that will probably cause node operators to manually disable bloom filters, leading to a vicious cycle where `p` rises much faster than expected, so that bloom filters degrade to being only offered by some small quasi-centralised set of servers, likely run by the providers of wallets that use bloom filters. If bloom users don't/can't authenticate the peers they're getting bloom filters from, that's probably a pretty big hole in the security assumptions.

Even if that occurs, I think making bloom filters opt-in (or default enabled but only for whitelisted peers) is still the right call.

Agree with @MarcoFalke that calling it a command line change doesn't seem quite right.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by practicalswift** [comment]
Concept ACK

**Comment by laanwj** [comment]
> so that bloom filters degrade to being only offered by some small quasi-centralised set of servers, likely run by the providers of wallets that use bloom filters.

… after which they might as well switch to a protocol that is efficient for that use such as electrum (which doesn't really have any worse privacy expectations), and everyone, including users of that wallet software will be better off :blush:

File: doc/release-notes/release-notes-16152.md (hunk 3 above)
Comment (TheBlueMatt) [reply doc/release-notes/release-notes-16152.md]: I dont know what the list of categories are, but I think a default parameter change is a CLI change. Happy to change to whatever.

**Comment by jnewbery** [comment]
Release notes fixed here: https://github.com/jnewbery/bitcoin/tree/pr16152_release_notes

@TheBlueMatt - I think if you reset to that branch, this should be mergeable, since the code changes have already been ACKed by @ajtowns , @jnewbery , @MarcoFalke, @luke-jr, @jtimon and @laanwj, and the release notes will be re-reviewed at release time.

**Comment by MarcoFalke** [comment]
As mentioned earlier (https://github.com/bitcoin/bitcoin/pull/16152#issuecomment-499202044), this probably requires a very brief post to the mailing list, that explains the change in p2p network behavior and mentions how long wallet developers have until their software breaks.

**Comment by luke-jr** [comment]
It won't break. People will continue to use old versions for years into the future. Some might turn NODE_BLOOM back on, too.

**Comment by MarcoFalke** [comment]
Running versions of Bitcoin Core that are EOL is not supported. So if there was a DOS vector against them, it wouldn't be fixed and the nodes could be shut down remotely. After those are shut down, the ones that have it turned on explicitly might also die off because they are flooded with filters at that point. So effectively after 0.18 is EOL, this p2p feature should be considered dead. Implying that it will be available forever is not helpful. I'd rather put a clear date on its expiry to give 3rd party developers a timescale to fix things. Otherwise they will rely on it never dying off and keep shipping software that might break at any point in time.

**Comment by jnewbery** [comment]
> I'd rather put a clear date on its expiry to give 3rd party developers a timescale to fix things.

I agree. Much better to overcommunicate to the ecosystem through the mailing list than undercommunicate.

**Comment by luke-jr** [comment]
Sure, no harm notifying people. My point was simply that 0.19 isn't suddenly a cut off. It will disappear gradually at most. And this only disables it by default - it's still there if someone wants to enable it (eg, for their own usage). It would even make sense to enable it for trusted peers by default, until a better solution exists. So it's not really dead at all yet.

**Comment by ajtowns** [comment]
ACK 2518c0643ba99f7da80bfe24aaad0921184ead78 ; just changes the release notes which now look good to me. Unsurprisingly, code still looks good, still compiles and tests pass for me.

**Comment by TheBlueMatt** [comment]
Tweaked the release notes to use the "P2P" header (thanks @jnewbery for answering my question of what the proper header is, happy to credit you in the commit if you want!), and rewrote the text to indicate that we don't anticipate it being a fast process (if anything, I'm happy to run a few of these, they aren't *that* bad if you have a super beefy machine). Happy to send a mail to the ML with the release notes text if this gets merged.

**Comment by jnewbery** [comment]
ACK bead32e31e399090af30b2ee3539995d4105a66d

The only difference from previous branch is updated release notes, which look good to me.

Please also consider this an ongoing ACK for _any_ release notes on top of f27309f55c4fa2b115525d72abb280757a568709 (which are the code changes)

> happy to credit you in the commit if you want

No need

**Review by kallewoof — APPROVED** [review]
ACK bead32e31e399090af30b2ee3539995d4105a66d

**Comment by NicolasDorier** [comment]
ACK

**Comment by MarcoFalke** [comment]
> Happy to send a mail to the ML with the release notes text if this gets merged.

Yes, please

File: doc/release-notes/release-notes-16152.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+P2P Changes
+-----------
+- The default value for the -peerbloomfilters configuration option (and, thus, NODE_BLOOM support) has been changed to false.
+  This resolves well-known DoS vectors in Bitcoin Core, especially for nodes with spinning disks. It is not anticipated that
+  this will result in a significant lack of availability of NODE_BLOOM-enabled nodes in the coming years, however, clients
+  which rely on the availability of NODE_BLOOM-supporting nodes on the P2P network should consider the process of migrating
+  to a more modern (and less trustful and privacy-violating) alternative over the coming years.
```
Comment (luke-jr) [inline doc/release-notes/release-notes-16152.md]: No such alternative exists...

**Comment by ShadowOfHarbringer** [comment]
ACK

**Comment by SmartArray** [comment]
I fully understand the purpose of having bloom filters disabled by default.
Are we trying to serve an alternative? Like this one: https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki

**Comment by runvnc** [comment]
Bloom filters are the way that lightweight wallets operate without centralized servers.

There is no evidence of a significant DoS issue.

This is a clear effort to sabotage Bitcoin.

**Comment by cculianu** [comment]
What's the matter? Liquid network is not the smash hit Blockstream hoped it would be and you need to do this to cripple SPV wallets? Or what?

**Pull request merged**